pcsx2/DEV9/AdapterUtils.cpp

```diff
@@ -244,30 +246,93 @@ std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
 
 	return std::nullopt;
 }
-#elif defined(__POSIX__)
-#ifdef __linux__
+#else
 std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
 {
-	struct ifreq ifr;
-	strcpy(ifr.ifr_name, adapter->ifa_name);
-
-	int fd = socket(AF_INET, SOCK_DGRAM, 0);
-	int ret = ioctl(fd, SIOCGIFHWADDR, &ifr);
-	close(fd);
+	int sd;
+	struct ifreq ifr, *ifrp;
+	struct ifconf ifc;
+	char buf[1024] = {0};
+	int n, i;
+	MAC_Address macAddr;
+#ifdef AF_LINK
+	struct sockaddr_dl* sdlp;
+#endif
 
-	if (ret == 0)
-		return *(MAC_Address*)ifr.ifr_hwaddr.sa_data;
+/*
+ * BSD 4.4 defines the size of an ifreq to be
+ * max(sizeof(ifreq), sizeof(ifreq.ifr_name)+ifreq.ifr_addr.sa_len
+ * However, under earlier systems, sa_len isn't present, so the size is 
+ * just sizeof(struct ifreq)
+ */
+#ifdef HAVE_SA_LEN
+#ifndef max
+#define max(a, b) ((a) > (b) ? (a) : (b))
+#endif
+#define ifreq_size(i) max(sizeof(struct ifreq), \
+	sizeof((i).ifr_name) + (i).ifr_addr.sa_len)
+#else
+#define ifreq_size(i) sizeof(struct ifreq)
+#endif /* HAVE_SA_LEN*/
 
-	return std::nullopt;
-}
+	sd = socket(AF_INET, SOCK_DGRAM, IPPROTO_IP);
+	if (sd < 0)
+	{
+		return std::nullopt;
+	}
+	memset(buf, 0, sizeof(buf));
+	ifc.ifc_len = sizeof(buf);
+	ifc.ifc_buf = buf;
+	if (ioctl(sd, SIOCGIFCONF, (char*)&ifc) < 0)
+	{
+		close(sd);
+		return std::nullopt;
+	}
+	n = ifc.ifc_len;
+	for (int i = 0; i < ifc.ifc_len; i += ifreq_size(*ifrp))
+	{
+		ifrp = ((struct ifreq*)(ifc.ifc_buf + i));
+		if (strcmp(ifrp->ifr_name, adapter->ifa_name) != 0)
+			continue;
+		strncpy(ifr.ifr_name, ifrp->ifr_name, IFNAMSIZ);
+#ifdef SIOCGIFHWADDR
+		if (ioctl(sd, SIOCGIFHWADDR, &ifr) < 0)
+			continue;
+		memcpy(&macAddr, &ifr.ifr_hwaddr.sa_data, 6);
 #else
-std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
-{
-	Console.Error("DEV9: Unsupported OS, can't get MAC address");
+#ifdef SIOCGENADDR
+		if (ioctl(sd, SIOCGENADDR, &ifr) < 0)
+			continue;
+		memcpy(&macAddr, &ifr.ifr_enaddr, 6);
+#else
+#ifdef AF_LINK
+		sdlp = (struct sockaddr_dl*)&ifrp->ifr_addr;
```

Comment:
Still technically type punning, but shrug, there's enough of it in PCSX2 that would break if it didn't work as expected.

Strictly speaking though, if you wanted to avoid it, I believe you'd have to use a union.

Reply:
Could you give me some hints about how to do it? I don't understand how a `union` can solve the casting issue...

Reply:
See https://github.com/PCSX2/pcsx2/blob/master/pcsx2/DEV9/PacketReader/IP/IP_Address.h as an example

However, having only recently looked into this, isn't using a union to type pun still undefined behaviour in C++?

Reply:
Seems like it may be. I guess you'd have to memcpy it then :/